patches/server/0949-Paper-Plugins.patch

```diff
@@ -0,0 +1,6590 @@
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/manager/MultiRuntimePluginProviderStorage.java
+@@ -0,0 +1,48 @@
++package io.papermc.paper.plugin.manager;
++
++import com.mojang.logging.LogUtils;
++import io.papermc.paper.plugin.entrypoint.Entrypoint;
++import io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler;
++import io.papermc.paper.plugin.provider.PluginProvider;
++import io.papermc.paper.plugin.provider.type.paper.PaperPluginParent;
++import io.papermc.paper.plugin.storage.ServerPluginProviderStorage;
++import org.bukkit.plugin.java.JavaPlugin;
++
++import java.util.ArrayList;
++import java.util.List;
++
++public class MultiRuntimePluginProviderStorage extends ServerPluginProviderStorage {
++
++    private static final org.slf4j.Logger LOGGER = LogUtils.getLogger();
++    private final List<JavaPlugin> provided = new ArrayList<>();
++
++    @Override
++    public void register(PluginProvider<JavaPlugin> provider) {
++        if (provider instanceof PaperPluginParent.PaperServerPluginProvider) {
++            LOGGER.warn("Skipping loading of paper plugin requested from SimplePluginManager.");
++            return;
++        }
++        super.register(provider);
++        /*
++        Register the provider into the server entrypoint, this allows it to show in /plugins correctly. Generally it might be better in the future to make a separate storage,
++         as putting it into the entrypoint handlers doesn't make much sense.
++         */
++        LaunchEntryPointHandler.INSTANCE.register(Entrypoint.PLUGIN, provider);
++    }
++
++    @Override
++    public void processProvided(JavaPlugin provided) {
++        super.processProvided(provided);
++        this.provided.add(provided);
++    }
++
++    @Override
++    public boolean exitOnCycleDependencies() {
++        return false;
++    }
++
++    public List<JavaPlugin> getLoaded() {
++        return this.provided;
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/manager/NormalPaperPermissionManager.java b/src/main/java/io/papermc/paper/plugin/manager/NormalPaperPermissionManager.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..29f8845bbad88e88e11e6365a867e37228bcf6fd
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/manager/NormalPaperPermissionManager.java
+@@ -0,0 +1,39 @@
++package io.papermc.paper.plugin.manager;
++
++import org.bukkit.permissions.Permissible;
++import org.bukkit.permissions.Permission;
++
++import java.util.*;
++
++class NormalPaperPermissionManager extends PaperPermissionManager {
++
++    private final Map<String, Permission> permissions = new HashMap<>();
++    private final Map<Boolean, Set<Permission>> defaultPerms = new LinkedHashMap<>();
++    private final Map<String, Map<Permissible, Boolean>> permSubs = new HashMap<>();
++    private final Map<Boolean, Map<Permissible, Boolean>> defSubs = new HashMap<>();
```

Comment:
Maps of booleans are pretty scuffed, idk if it's better to change the abstract classwith a little bit of extra diff for the Spigot impl